pkg/compare/container.go

```diff
@@ -0,0 +1,140 @@
+	f := func(c rune) bool {
+		return c == ':'
+	}
+	sections := strings.FieldsFunc(path, f)
+
+	if len(sections) == 3 {
+		image := sections[0] + ":" + sections[1]
+		referencePath := sections[2]
+		return image, referencePath, nil
+	}
+	return "", "", fmt.Errorf("incorrect path passed to -r, it must follow this format: container://<IMAGE>:<TAG>:/path/to/metadata.yaml")
+}
+
+// Use var's so that we can mock functions in tests.
+var execCommand = exec.Command
+var lookPath = exec.LookPath
+
+// runEngineCommand runs a podman/docker command with sudo if necessary.
+// Returns the stdout (out) and stderr (err) of the command.
+func (engine *engine) runEngineCommand(args ...string) ([]byte, error) {
+	var out []byte
+	var err error
+	if engine.requiresSudo {
+		args = append([]string{engine.name}, args...) // Prepend engine name to args
+		out, err = execCommand("sudo", args...).CombinedOutput()
+	} else {
+		out, err = execCommand(engine.name, args...).CombinedOutput()
+	}
+
+	return out, err
+}
+
+// newEngine checks if Podman or Docker are in the system's PATH, and returns an engine with a name and a boolean
+// that indicates if sudo is needed for future commands. Prefers Podman. Returns an error if neither engine is found.
+var newEngine = func() (*engine, error) {
+	if _, err := lookPath("podman"); err == nil {
+		return &engine{name: "podman", requiresSudo: false}, nil
+	}
+
+	if _, err := lookPath("docker"); err == nil {
+		_, err = execCommand("docker", "images").Output() // If this errors out, we need to use sudo, return true.
+		return &engine{name: "docker", requiresSudo: err != nil}, nil
+	}
+	return &engine{name: "", requiresSudo: false}, fmt.Errorf("you do not have Podman or Docker on your PATH")
+}
+
+// pullContainer pulls an image, runs it using the provided engine, and stores the corresponding containerID in the engine struct
+func (engine *engine) pullAndRunContainer(image string) error {
+	// run because copy requires a running or stopped container
+	// -d to output container ID
+	out, err := engine.runEngineCommand("run", "-d", image)
+	if err != nil {
+		return fmt.Errorf("could not pull/run container: %s", out)
+	}
+	engine.containerID = strings.TrimSpace(string(out)) // Convert bytes to string and trim new line
+	return nil
+}
+
+// extractReferences copies the directory in the container that contains the reference configs into a temporary directory,
+// and stores the path to the new directory in the engine struct.
+func (engine *engine) extractReferences(pathToMetadata string, dname string) error {
+
+	out, err := engine.runEngineCommand("cp", engine.containerID+":"+pathToMetadata, dname)
+	if err != nil {
+		return fmt.Errorf("could not copy templates from container: %s", out)
+	}
+	engine.tempDir = filepath.Join(dname, filepath.Base(pathToMetadata))
+	return nil
+}
+
+// cleanup stops and removes the container used to extract the reference configs.
+func (engine *engine) cleanup() error {
```

Comment:
This function can only return `nil` so maybe just
```suggestion
func (engine *engine) cleanup() {
```